swap/cashout_test.go

```diff
@@ -0,0 +1,196 @@
+// (at your option) any later version.
+//
+// The Swarm library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the Swarm library. If not, see <http://www.gnu.org/licenses/>.
+
+package swap
+
+import (
+	"context"
+	"crypto/ecdsa"
+	"math/big"
+	"testing"
+
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/log"
+	contract "github.com/ethersphere/swarm/contracts/swap"
+)
+
+func newSignedTestCheque(testChequeContract common.Address, beneficiaryAddress common.Address, cumulativePayout *big.Int, signingKey *ecdsa.PrivateKey) (*Cheque, error) {
+	cheque := &Cheque{
+		ChequeParams: ChequeParams{
+			Contract:         testChequeContract,
+			CumulativePayout: cumulativePayout.Uint64(),
+			Beneficiary:      beneficiaryAddress,
+		},
+		Honey: cumulativePayout.Uint64(),
+	}
+
+	sig, err := cheque.Sign(signingKey)
+	if err != nil {
+		return nil, err
+	}
+	cheque.Signature = sig
+	return cheque, nil
+}
+
+// TestContractIntegration tests a end-to-end cheque interaction.
+// First a simulated backend is created, then we deploy the issuer's swap contract.
+// We issue a test cheque with the beneficiary address and on the issuer's contract,
+// and immediately try to cash-in the cheque
+// afterwards it attempts to cash-in a bouncing cheque
+func TestContractIntegration(t *testing.T) {
+	backend := newTestBackend(t)
+	reset := setupContractTest()
+	defer reset()
+
+	payout := big.NewInt(42)
+
+	chequebook, err := testDeployWithPrivateKey(context.Background(), backend, ownerKey, ownerAddress, payout)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	cheque, err := newSignedTestCheque(chequebook.ContractParams().ContractAddress, beneficiaryAddress, payout, ownerKey)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	opts := bind.NewKeyedTransactor(beneficiaryKey)
+
+	tx, err := chequebook.CashChequeBeneficiaryStart(opts, beneficiaryAddress, payout, cheque.Signature)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	receipt, err := contract.WaitForTransactionByHash(context.Background(), backend, tx.Hash())
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	cashResult := chequebook.CashChequeBeneficiaryResult(receipt)
+	if receipt.Status != 1 {
+		t.Fatalf("Bad status %d", receipt.Status)
+	}
+	if cashResult.Bounced {
+		t.Fatal("cashing bounced")
+	}
+
+	// check state, check that cheque is indeed there
+	result, err := chequebook.PaidOut(nil, beneficiaryAddress)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if result.Uint64() != cheque.CumulativePayout {
+		t.Fatalf("Wrong cumulative payout %d", result)
+	}
+	log.Debug("cheques result", "result", result)
+
+	// create a cheque that will bounce
```

Comment:
You mean a separate test? That would probably make sense. Not sure why those two things are in the same test in the first place but I didn't want to rewrite the tests here too much since the pr is about something else.

Reply:
Having positive and negative tests in the same run can reduce risks of false positive, i.e. bounced because of an edge case where the test was setup badly or special conditions. If we first have a positive result, which means "it worked in normal operation", and then a negative test in the same run, we can reduce risk of test success due to coincidence. So personally I am not opposed to keep it in the same test, but I would support to split too, or to split and call the negative from here too.